docs/decisions/0012-assignment-based-policies.rst

```diff
@@ -0,0 +1,133 @@
+may *allocate* new assignments, read or modify their state, and determine
+correct aggregate facts about the amount of value currently *allocated* or *spent*
+against an assignment-based ``SubsidyAccessPolicy``.
+
+
+New ``content_assignments`` Django app
+--------------------------------------
+We built a new ``content_assignments`` Django application.  It persists data about
+allocated assignments - these assignments, stored as ``LearnerContentAssignment`` model instances,
+represent the fact that an admin assigned some course to a learner at a given time and price.
+The ``subsidy_access_policy`` app’s business logic will integrate with this new app.
+A model called ``content_assignments.AssignmentConfiguration`` is depended upon by the ``LearnerContentAssignment``
+model and the base ``SubsidyAccessPolicy`` model via a foriegn key relationship.
+
+The default, happy-state of a ``LearnerContentAssignment`` is ``allocated``.  We talk about
+"allocation" as the act of creating a new assignment within a given policy (or "budget").
+Allocations should be taken into account when determining current spend for an
+assignment-based policy/budget (see section below).
+
+This new Django app will introduce asynchronous celery tasks to:
+
+- Link pending, allocated learners to an enterprise
+- Notify learners of new allocated assignments, or to remind them of existing allocations.
+
+New assignment-based policy type
+--------------------------------
+The ``SubsidyAccessPolicy`` model and business logic will now depend on the
+state of assignment records. This aligns with the policy domain's role as an
+"octopus enforcer" - it depends on all of the other domains of our software
+that must be considered in computing queries and commands about access to content
+covered by a policy.
+
+We'll introduce a new type of assignment-based policy model, with two new, top-level methods
+that allow for querying and commanding about allocation of new assignment records:
+
+- ``can_allocate`` takes allocated ``LearnerContentAssignment`` records related to this policy
+  into account to determine if some number of learners can have assignment records
+  allocated in this policy for a given content key  and its current price.
+- ``allocate`` is the command through which new assignment records should
+  be allocated against a given policy.  It queries ``can_allocate`` before
+  allocation occurs.
+
+Before determining if new assignments can be allocated for a set of learners,
+the assignment-based policy logic must now also take into account, via the sum of:
+
+- The total cost of new assignments to be allocated.
+- The total cost of all existing allocated (but not accepted, failed, or cancelled) assignments related to the policy.
+- Must not exceed the remaining balance of the policy’s subsidy.
+- Must not exceed the remaining spend limit configured for the policy.
+
+Furthermore, we'll have to modify *redemption* logic for assignment-based policies as follows:
```

Comment:
While not _redemption_ logic per se, should we mention any implications to the `credits_available` API? For example, given the `credits_available` API returns all Budgets (policies) for a given customer/learner that are redeemable (agnostic to a piece of content), under what situations would the new assignment policy type be returned by `credits_available` API? E.g., only when an assignment(s) have been accepted/redeemed by the learner or when there is an allocated assignment?

Reply:
Yeah, good point.  I'd say as long as an assignment-based policy is active, current, and has balance, `credits_available` probably should return it (as long as the learner is linked to the enterprise).

Reply:
Makes sense to me, so long as the product/UX requirements expects the "Learner Credit" subsidy box to appear on the Learner Portal dashboard page and for the catalog tied to the assignment policy to be included in the Learner Portal search results when the learner hasn't been assigned any content yet.

Reply:
Actually, maybe that's not the point of `credits_available`?  Is the purpose to inform the learner portal of budgets with which the learner could redeem *right now*?  If so, then maybe it should only describe budgets with allocated assignments, as you suggested.

Reply:
- ``credits_available`` This endpoint returns a list of current, active policies
  with credit availble to be redeemed **immediately**.  Therefore, we'll have to conditionally
  include assignment-based policies in this endpoint's response payload - if the requesting
  learner has an allocated assignment record associated with a current, active policy, that
  policy record should be included in the response payload.